### ISSUE TITLE
The keras.layers.CategoricalEncoding layer removes batch dimension for single-element batches.

### ISSUE THREAD

**Issue by foxik** [keras-team/keras#15900]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): **No**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Debian stable**
- TensorFlow installed from (source or binary): **binary**
- TensorFlow version (use command below): **2.7, tf-nightly**
- Python version: **3.7**
- Bazel version (if compiling from source):
- GPU model and memory:
- Exact command to reproduce:

**Describe the problem**.
When `tf.keras.layers.CategoryEncoding(output_mode="one_hot")` is used on a single-element batch with a scalar value (i.e., input with shape `[1]`), it removes the batch dimension on the output, i.e., creates an output with shape `[num_classes]`. When an input with shape `[2]` is used, the output has shape `[2, num_classes]`.

Therefore, if the `CategoryEncoding` is used in a model passing scalar-per-batch-example through it, the model breaks when used on a single-element batch.

**Describe the current behavior**.
The code
```
category_encoding = tf.keras.layers.CategoryEncoding(16, output_mode="one_hot")
for input_shape in ([1], [2], [3]):
  print("For input shape {} the CategoryEncoding produces output with shape {}".format(
      input_shape, category_encoding(tf.zeros(input_shape, dtype=tf.int32)).shape))
```
produces an output
```
For input shape [1] the CategoryEncoding produces output with shape (16,)
For input shape [2] the CategoryEncoding produces output with shape (2, 16)
For input shape [3] the CategoryEncoding produces output with shape (3, 16)
```

**Describe the expected behavior**.
I would expect the output of the above snippet to be
```
For input shape [1] the CategoryEncoding produces output with shape (1, 16)
For input shape [2] the CategoryEncoding produces output with shape (2, 16)
For input shape [3] the CategoryEncoding produces output with shape (3, 16)
```

**Further Information**

Note that the documentation at https://www.tensorflow.org/api_docs/python/tf/keras/layers/CategoryEncoding states that
> "one_hot": Encodes each individual element in the input into an array of num_tokens size, containing a 1 at the element index. If the last dimension is size 1, will encode on that dimension. If the last dimension is not size 1, will append a new dimension for the encoded output.

So the described "remove batch-dimension if of size 1" issue can be seen as an application of this "remove the last dimension if of size 1" rule. However, I would argue that the "remove the last dimension if of size 1" should be only applied to **non-batch** dimensions.

If this is deemed to be a bug, the documentation has to be updated.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): **no**

**Standalone code to reproduce the issue**.

https://colab.research.google.com/drive/1cXGc4bN0Cg660-oV3B7omvW9Sp8ebav2?usp=sharing

**Comment by tilakrayal** [comment]
@jvishnuvardhan ,
I was able to reproduce the issue in v2.7 and nightly.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/6451be832f2bcc872d06898773b276e2/categoryencoding.ipynb).

**Comment by gadagashwini** [comment]
@foxik,
Looks like its expected behaviour.
**input with shape [2] ==> output shape [2, num_classes]**
```
import tensorflow as tf

category_encoding = tf.keras.layers.CategoryEncoding(num_tokens=4, output_mode="one_hot")
category_encoding(tf.zeros([2],dtype=tf.int32))
```
```
<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[1., 0., 0., 0.],
       [1., 0., 0., 0.]], dtype=float32)>
```
**input with shape [1] ==> output shape [num_classes,]** ==> `numpy=array([1., 0., 0., 0.]`
1-D Tensor ==> output shape ==> shape=(number_elements,)
```
import tensorflow as tf

category_encoding = tf.keras.layers.CategoryEncoding(num_tokens=4, output_mode="one_hot")
category_encoding(tf.zeros([1],dtype=tf.int32))
```

`<tf.Tensor: shape=(4,), dtype=float32, numpy=array([1., 0., 0., 0.], dtype=float32)>`

**Example of 1-D Tensor**
```
import tensorflow as tf
a = tf.constant([1, 2, 3, 4, 5, 6])
a
<tf.Tensor: shape=(6,), dtype=int32, numpy=array([1, 2, 3, 4, 5, 6], dtype=int32)>
```

**Comment by foxik** [comment]
Hi,

I understand that it can be seen as the application of the rule from the documentation (as stated above).

However, removing __batch dimension__ seems very wrong -- a valid model will break when executed with a batch size of 1 (while it will work with batch sizes 2 or more). So personally I think the batch dimension should never be removed (i.e., that the rule about removing dimensions of size 1 should apply only for all-but-the-first dimensions), and that the current behavior is actually a bug.

Cheers ;-)

**Comment by mattdangerw** [comment]
@foxik I think an issue with what you are proposing is we really can't assume a batch dimension at all. We need to support to flows generally speaking with preprocessing...

```
ds.map(preprocessing).batch(8)
ds.batch(8).map(preprocessing)
```

This makes our preprocessing as a whole much more flexible in terms of where it can be applied, particular when thinking about multi-worker settings. But it does lead to some ambiguous cases like this! Where we can't actually be sure at the layer level if a input with shape `[1]` is a batch or a single sample (and you would want to handle the output shape differently).

The best solution for dealing when dealing with batched scalars would probably be to add a second dimension to the input. A batch with one scalar element should have shape `[1, 1]`, a batch with two scalars should have shape `[2, 1]`, and the output will have shape `[bs, num_tokens]` in all cases. Would that work for you?

**Comment by foxik** [comment]
Hi,

thanks for the reply -- I did not think about using the layer in the `tf.data` pipeline, where single examples are processed (instead, I imagined the layer to be used only in the network, where the 0-th dimension is a batch dimension), so I understand why you want to keep the current behavior, even if it might cause problems in single-example batch inference.

We used the work-around with introducing an additional dimension ourselves, but it was surprising that we need to do it anyway.

Thanks for your answers, closing the issue. Cheers!
